### PR TITLE
Show failed point of lines in configuration file

### PR DESCRIPTION
**Pull request by matsune**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ ] I've updated the documentation if necessary.

### Motivation and Context

implement #11706 #11705

When Fastlane failed to parse configuration file (e.g. Fastfile, Matchfile) it should output lines around error point on console with error message.

### Description

↓Outputs message as below when caught NameError in Fastfile.

### CODE REVIEW

File: fastlane/lib/fastlane/fast_file.rb [hunk 1]
```diff
@@ -57,6 +57,8 @@ def parse(data, path = nil)
           # rubocop:enable Security/Eval
         rescue SyntaxError => ex
           line = ex.to_s.match(/#{Regexp.escape(relative_path)}:(\d+)/)[1]
+          UI.error("Error in your Fastfile at line #{line}")
```
Comment (janpio) [inline fastlane/lib/fastlane/fast_file.rb]: Isn't this duplicated 2 lines below?

Comment (matsune) [reply fastlane/lib/fastlane/fast_file.rb]: That's right. 😮

File: fastlane/lib/fastlane/lane_manager.rb [hunk 2]
```diff
@@ -55,6 +55,12 @@ def self.cruise_lane(platform, lane, parameters = nil, env = nil)
       e = nil
       begin
         ff.runner.execute(lane, platform, parameters)
+      rescue NameError => ex
+        print_lane_context
+        line = ex.backtrace.first.match("Fastfile:(\\d+):")[1]
```
Comment (joshdholtz) [inline fastlane/lib/fastlane/lane_manager.rb]: I ran this with a good `Fastfile` and a `Snapfile` with the contents of...
```rb
Upload
```
and I got an error on this line 🤔 

```sh
bundler: failed to load command: fastlane (/Users/josh/.gem/ruby/2.3.6/bin/fastlane)
NoMethodError: [!] undefined method `[]' for nil:NilClass
  /Users/josh/Projects/fastlane/fastlane/fastlane/lib/fastlane/lane_manager.rb:60:in `rescue in cruise_lane'
  /Users/josh/Projects/fastlane/fastlane/fastlane/lib/fastlane/lane_manager.rb:56:in `cruise_lane'
  /Users/josh/Projects/fastlane/fastlane/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
  /Users/josh/Projects/fastlane/fastlane/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
```

Any ideas? 😇

**Review by joshdholtz — CHANGES_REQUESTED** [review]

**Comment by matsune** [comment]
@joshdholtz I solved that problem, please recheck 🙏

**Comment by matsune** [comment]
@joshdholtz Thanks for reviewing!

It’s exactly as you said that non-Fastfile’s error should be pointed out its error line.
We can get error line from Exception’s backtrace when occurred `SyntaxError` or `NameError` at Fastfile.
e.g.) NameError at Fastfile
```ruby
7: lane :a do
8:   Upload
9: end
```
backtrace => `Fastfile:8:in 'block in parsing_binding’`

e.g.) SyntaxError at Fastfile
```ruby
7: lane :a do
8:   1+1)
9: end
```
backtrace => `[!] Syntax error in your Fastfile on line 8: Fastfile:8: syntax error, unexpected ')', expecting keyword_end`

And configuration files(e.g. Snapfile, Matchfile) also can get error line of `SyntaxError`.
e.g.) SyntaxError at Snapfile
```ruby
1:
2: 1+1)
3:
```
backtrace => `[!] Syntax error in your configuration file './fastlane/Snapfile' on line 2: (eval):2: syntax error, unexpected ')', expecting end-of-input`

However, configuration files cannot get error line from `NameError`.
e.g.) NameError at Snapfile
```ruby
1:
2: Upload
3:
```
backtrace => `NameError: [!] uninitialized constant FastlaneCore::ConfigurationFile::Upload  Did you mean?  UploadIO`

Configuration files parse its content using `eval`, but Exception `NameError` while `eval` is not containing error line in backtrace, so I think it’s too difficult to find line of NameError point.. 😢 
Thus I made a change to show error line at Fastfile’s SyntaxError and NameError, Configuration file’s SyntaxError.

**Review by joshdholtz — CHANGES_REQUESTED** [review]
@matsune Thanks for explaining that! That makes complete sense 👌 This looks super good. Are you able to write some tests are these? I think that should be the final thing 😊

**Comment by matsune** [comment]
@joshdholtz Sorry for late response. I added some tests for parsing syntax error fastfile.

**Review by joshdholtz — APPROVED** [review]
Thanks for writing the tests! Excited to have this PR go out 😊

File: fastlane/lib/fastlane/lane_manager_base.rb [hunk 3]
```diff
@@ -75,5 +75,14 @@ def self.print_lane_context
         rows: FastlaneCore::PrintTable.transform_output(rows)
       }))
     end
+
+    def self.print_error_line(ex)
+      error_line = ex.backtrace[0].match("Fastfile:(\\d+):")
```
Comment (jaleksynas) [inline fastlane/lib/fastlane/lane_manager_base.rb]: `backtrace[0]` is not always set for some reason and fastlane throws exception:

fastlane-2.96.1/fastlane/lib/fastlane/lane_manager_base.rb:80:in `print_error_line': [!] undefined method `match' for nil:NilClass (NoMethodError)

Null check?

File: fastlane_core/lib/fastlane_core/configuration/configuration_file.rb [hunk 4]
```diff
@@ -43,12 +43,14 @@ def initialize(config, path, block_for_missing, skip_printing_values = false)
 
       begin
         # rubocop:disable Security/Eval
-        eval(content) # this is okay in this case
+        puts(eval(content)) # this is okay in this case
```
Comment (tflhyl) [inline fastlane_core/lib/fastlane_core/configuration/configuration_file.rb]: could we remove this `puts`?

Comment (matsune) [reply fastlane_core/lib/fastlane_core/configuration/configuration_file.rb]: You may be right.
I had thought that `puts` is required to catch SyntaxError execption during `eval` … but I tried it now, it wasn’t necessary.

**Pull request merged**